### ISSUE TITLE
Option to not delete file when moving/copying it out of Play temp folder

### ISSUE THREAD

**Issue by mkurz** [playframework/playframework#8954]
Have a look at [this](https://github.com/playframework/playframework/pull/8889/files#r248240569) and [that](https://github.com/playframework/playframework/pull/8889/files#r248244196) comment.

If I use any of the `copyTo`, `moveTo` or `atomicMoveWithFallback` methods to copy/move a file **out** of the play temp folder, we always keep a "finalize" reference on the "new" file for each of these three methods:

https://github.com/playframework/playframework/blob/9a8b47007f17655001958414e77f5408d4ece137/framework/src/play/src/main/scala/play/api/libs/Files.scala#L104

https://github.com/playframework/playframework/blob/9a8b47007f17655001958414e77f5408d4ece137/framework/src/play/src/main/scala/play/api/libs/Files.scala#L136

https://github.com/playframework/playframework/blob/9a8b47007f17655001958414e77f5408d4ece137/framework/src/play/src/main/scala/play/api/libs/Files.scala#L173

meaning that if there is no reference anymore to the `TemporaryFile` (on which one of the above methods was called) the "new" file created (outside the play temp folder) by those methods will [be deleted](https://github.com/playframework/playframework/blob/9a8b47007f17655001958414e77f5408d4ece137/framework/src/play/src/main/scala/play/api/libs/Files.scala#L218-L221) when the garbage collector kicks in.

I am not sure if that is such a good behaviour. I mean, why would I move a file **out** of the play temp folder? Most likely to keep it permanent, or maybe to keep if for a while, process it (in another thread, another process/application, whatever) and then maybe delete it.

Therefore I suggest to add a param to *all* of these methods, something like `tmpFile.copyTo(..., removeOnFinalize = true)` (with a default that does not break compatibility).
All it would do is:
```scala
if (removeOnFinalize) {
    temporaryFileCreator.create(to)
}
```

Also actually I think the analysis in #8788 by @wsargent was wrong, #8788 has nothing to do with the Inode, but the fact that even newly moved or copied files (out of the play temp folder) will be deleted if their `TemporaryFile` instances gets garbage collected.

What do you think?
Or maybe I just get something completely wrong here?

**Comment by mkurz** [comment]
Also see https://github.com/playframework/playframework/issues/8788#issuecomment-454765760

**Comment by mkurz** [comment]
Other options how to fix this would be:
* Just never create/hold a "finalize" reference onto a moved or copied file
* Detect if we move/copy a file **out** of the Play temp folder, and if so, only then do not create/hold a "finalize" reference onto such a file

**Comment by mkurz** [comment]
Also look what the [docs say](https://www.playframework.com/documentation/2.6.x/ScalaFileUpload#Cleaning-up-temporary-files):
> ...and should be moved out of the temporary file system as soon as possible. Any temporary files that are not moved are deleted.

However this is not how it just works now, the GC will also cause to delete files which actually *do get moved out* of the "temporary file system"...

Actually I think the option
* Just never create/hold a "finalize" reference onto a moved or copied file

isn't so bad actually?
I mean if I call such a method I very likely will move the file out of the play temp folder (to keep it or whatever) - but I certainly don't want Play to remove that new file for me. Even if someone would move/copy a file *within* the play temp folder (even this wouldn't make lots of sense), which (with this approach) would result in not having a finalize reference to it, the "new" file would be deleted by the reaper at some point. I think that would be totally ok.

**Comment by wsargent** [comment]
Oh good grief.  You're right, it's right there in the API:

```
def copyTo(to: Path, replace: Boolean): TemporaryFile 
def moveTo(to: java.io.File, replace: Boolean = false): TemporaryFile
```

A TemporaryFile is temporary, and should be deleted.  I think this should return `File`.

**Comment by mkurz** [comment]
I will provide a pull request for the 2.6.x branch which does not break compatibility.

**Comment by wsargent** [comment]
s/File/Path/

**Comment by mkurz** [comment]
Fix is here #8958

**Comment by mkurz** [comment]
Fixed via #8958
